src/decoder/flac.rs

```diff
@@ -145,27 +611,130 @@ where
             self.current_block_off = 0;
             let buffer = mem::take(&mut self.current_block);
-            match self.reader.blocks().read_next_or_eof(buffer) {
+            match self
+                .reader
+                .as_mut()
+                .expect("reader should always be Some")
+                .blocks()
+                .read_next_or_eof(buffer)
+            {
                 Ok(Some(block)) => {
                     self.current_block_channel_len = (block.len() / block.channels()) as usize;
                     self.current_block = block.into_buffer();
                 }
-                _ => return None,
+                Ok(None) | Err(_) => {
+                    // No more blocks or error, current_block becomes empty
+                    // (buffer was consumed by read_next_or_eof)
+                    return None;
+                }
             }
         }
     }
+
+    /// Returns bounds on the remaining length of the iterator.
+    ///
+    /// Provides accurate size estimates based on FLAC metadata when available.
+    /// This information can be used by consumers for buffer pre-allocation
+    /// and progress indication.
+    ///
+    /// # Returns
+    ///
+    /// A tuple `(lower_bound, upper_bound)` where:
+    /// - `lower_bound`: Minimum number of samples guaranteed to be available
+    /// - `upper_bound`: Maximum number of samples that might be available (None if unknown)
+    ///
+    /// # Accuracy
+    ///
+    /// - **With metadata**: Exact remaining sample count (lower == upper)
+    /// - **Without metadata**: Conservative estimate based on current block
+    /// - **Stream exhausted**: (0, Some(0))
+    ///
+    /// # Implementation
+    ///
+    /// The lower bound counts buffered samples that are immediately available.
+    /// The upper bound uses total stream metadata when available for precise counting.
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        // Samples already decoded and buffered (guaranteed available)
+        let buffered_samples = self
+            .current_span_len()
+            .unwrap_or(0)
+            .saturating_sub(self.current_block_off);
+
+        if let Some(total_samples) = self.total_samples {
+            let total_remaining = total_samples.saturating_sub(self.samples_read) as usize;
+            (buffered_samples, Some(total_remaining))
+        } else if self.current_block.is_empty() {
+            // Stream exhausted (no more blocks available)
+            (0, Some(0))
+        } else {
+            (buffered_samples, None)
+        }
+    }
 }
 
-/// Returns true if the stream contains FLAC data, then tries to rewind it to where it was.
-fn is_flac<R>(mut data: R) -> bool
+/// Probes input data to detect FLAC format.
+///
+/// This function attempts to parse the FLAC magic bytes and stream info header to determine if the
+/// data contains a valid FLAC stream. The stream position is restored regardless of the result.
+///
+/// # Arguments
```

Comment:
We have a type system for this info. I'd propose we not duplicate the type information in comments that will rot away over time (and add many many lines)